### PR TITLE
gh-112346: Bugfix: Remove faster codepath from gzip.compress as it introduces behavioral inconsistencies

### PR DESCRIPTION
**Pull request by rhpvorderman**
As mentioned in the linked issue, delegating to `zlib.compress` causes behavorial changes in the OS byte. As a result this change caused reproducible build failures. The separate codepath caused issues before as well:
https://github.com/python/cpython/issues/90425

Therefore the codepath should be removed. Speed ain't everything. The mention of zlib.compress in the documentation is maintained so users who not have specific requirements can use the speedier path. 

This change needs to be backported to 3.11 and 3.12

* Issue: gh-112346

----
📚 Documentation preview 📚: https://cpython-previews--114116.org.readthedocs.build/

### CODE REVIEW

**Comment by rhpvorderman** [comment]
So is there any procedure for News entries for bugfixes that need to be backported to older versions of CPython?

**Comment by iii-i** [comment]
I would suggest the following fixup to make this work on BE (mtime is always LE, regardless of the host endianness):

```
--- a/Lib/test/test_gzip.py
+++ b/Lib/test/test_gzip.py
@@ -725,7 +725,7 @@ def test_issue112346(self):
         for mtime in (0, 42):
             with self.subTest(mtime=mtime):
                 compress = gzip.compress(data1, compresslevel=1, mtime=mtime)
-                assert struct.unpack("IxB", compress[4:10]) == (mtime, 255)
+                assert struct.unpack("<IxB", compress[4:10]) == (mtime, 255)
 
     def test_decompress(self):
         for data in (data1, data2):
```

**Comment by rhpvorderman** [comment]
Thanks for the suggestion @iii-i . That was an oversight on my part. I am going to pin in the hope this fix gets noticed.

**Comment by rhpvorderman** [comment]
ping

**Comment by rhpvorderman** [comment]
Ping @gpshead . Could you shine your light on this? It's only a very small change and it has remained under the radar for almost half a year now.

File: Lib/test/test_gzip.py [hunk 1]
```diff
@@ -554,14 +554,21 @@ def test_compress_mtime(self):
...
+                                                 mtime=mtime)
                 self.assertIn(data1, nocompress)
                 self.assertNotIn(data1, yescompress)
 
+    def test_issue112346(self):
+        # The OS byte should be 255, this should not change between Python versions.
+        for mtime in (0, 42):
+            with self.subTest(mtime=mtime):
+                compress = gzip.compress(data1, compresslevel=1, mtime=mtime)
+                assert struct.unpack("<IxB", compress[4:10]) == (mtime, 255)
```
Comment (serhiy-storchaka) [inline Lib/test/test_gzip.py]: Use TestCase assertion methods instead of the assert statement.

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM, except using `assert` in the test.

File: Lib/test/test_gzip.py (hunk 1 above)
Comment (rhpvorderman) [reply Lib/test/test_gzip.py]: Done!

**Comment by rhpvorderman** [comment]
Thanks for reviewing this so quickly @serhiy-storchaka . I added a news entry.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by gpshead** [comment]
I suggest not making this change given how long it has been the behavior of released Pythons.  We cannot backport this to 3.11 as it is not a security issue.  And anyone picking a change like this up in a patch release is just as likely to run into problems with the behavior change between differing Python versions.

**Comment by rhpvorderman** [comment]
@gpshead That makes sense. Sorry, I went into panic mode when code changes I made gave someone some reproducibility troubles. 
I do hope that the two separate code paths do not cause any more in the future.

**Pull request closed without merging**